### PR TITLE
Support JuMP 1.2.0

### PR DESCRIPTION
**Pull request by abelsiqueira**
cc. @amontoison @odow 

Trying to prepare for JuMP 1.2.0 (#116).

### CODE REVIEW

**Comment by abelsiqueira** [comment]
I am stopping for the day. The NLS part (`user_operators` and `subexpressions _as_julia_expressions`) are untouched.

File: test/nls_problems/nlshs20.jl [hunk 1]
```diff
@@ -4,11 +4,11 @@ function nlshs20()
   lvar = [-0.5; -Inf]
   uvar = [0.5; Inf]
   x0 = [-2.0; 1.0]
-  @variable(model, lvar[i] ≤ x[i = 1:2] ≤ uvar[i], start = x0[i])
+  @variable(model, lvar[i] <= x[i = 1:2] <= uvar[i], start = x0[i])
 
-  @NLconstraint(model, x[1] + x[2]^2 ≥ 0)
-  @NLconstraint(model, x[1]^2 + x[2] ≥ 0)
-  @NLconstraint(model, x[1]^2 + x[2]^2 ≥ 1)
```
[the remark this answers is not in the document]
Comment (odow) [reply test/nls_problems/nlshs20.jl:11]: Fixed on the latest commit. You can revert these changes.

File: src/utils.jl [hunk 2]
```diff
@@ -384,29 +384,30 @@ function parser_nonlinear_expression(cmodel, nvar, F; hessian::Bool = true)
     @variable(Fmodel, x[1:nvar])
     JuMP._init_NLP(Fmodel)
     @objective(Fmodel, Min, 0.0)
-    Fmodel.nlp_data.user_operators = cmodel.nlp_data.user_operators
```
Comment (odow) [inline src/utils.jl:387]: I had a brief look at this code, and I got very confused as to what you are actually trying to do.

Isn't the goal just to get an evaluator for a JuMP model? Why build a new model for each expression?

File: src/moi_nlp_model.jl [hunk 3]
```diff
@@ -18,28 +22,47 @@ Construct a `MathOptNLPModel` from a `JuMP` model.
...
   nvar, lvar, uvar, x0 = parser_JuMP(jmodel)
 
+  eval = NLPEvaluator(jmodel)
   nnln = num_nonlinear_constraints(jmodel)
 
-  nl_lcon = nnln == 0 ? Float64[] : map(nl_con -> nl_con.lb, jmodel.nlp_data.nlconstr)
-  nl_ucon = nnln == 0 ? Float64[] : map(nl_con -> nl_con.ub, jmodel.nlp_data.nlconstr)
+  nl_lcon = fill(-Inf, nnln)
+  nl_ucon = fill(Inf, nnln)
+  for (i, (_, constraint)) in enumerate(eval.model.constraints)
```
[the remark this answers is not in the document]
Comment (abelsiqueira) [reply src/moi_nlp_model.jl:30]: Thanks

File: src/utils.jl (hunk 2 above)
Comment (abelsiqueira) [reply src/utils.jl:387]: This was initially [this post](https://discourse.julialang.org/t/jump-creating-a-nonlinear-least-squares-model/10706), and the current version is how @amontoison made it work with MathOptInterface. The `user_operators` come from https://github.com/JuliaSmoothOptimizers/NLSProblems.jl/issues/14#issuecomment-393025755 to get the user-defined functions.
The original idea is that we take the expressions of `cmodel` but we have to swap the variables of `Fmodel` and add it as constraints to Fmodel.

**Comment by abelsiqueira** [comment]
Thanks for the comments @odow, I've replied to them.
One more thing, I saw you mentioned `MOI.get(model, MOI.NLPBlock())`, but this returned `nothing`:

```julia
model = Model()
@variable(model, x[1:3])
@NLobjective(model, Min, sum(x[i]^2 for i = 1:3))
@NLconstraint(model, x[1] * x[2] * x[3] == 1)
@constraint(model, x[1] + x[2] + x[3] == 999)
@NLconstraint(model, x[1] * x[2] >= 2)
@NLconstraint(model, x[1] * x[3] <= 3)
@NLconstraint(model, 4 <= x[2] * x[3] <= 5)
MOI.get(model, MOI.NLPBlock())
```

**Comment by abelsiqueira** [comment]
Ok, CI passed, but there are a few things that I know should be done and maybe other issues I can't see at the moment.
TODO:
- Use sparse structures that are computed on the `MathOptNL?Model` function
- Clean up
- Try to avoid using internal structures

**Comment by amontoison** [comment]
@abelsiqueira @tmigot 
I updated the PR but I have an issue with the nonlinear part of the Jacobian. It seems to come from NLPModels.jl: https://github.com/JuliaSmoothOptimizers/NLPModels.jl/blob/main/src/nlp/api.jl#L193-L195

**Comment by tmigot** [comment]
There seems to be an error with `hs70`:
```
ERROR: LoadError: BoundsError: attempt to access 0-element Vector{Bool} at index [1]
Stacktrace:
  [1] getindex
    @ ./array.jl:924 [inlined]
  [2] _topological_sort(starts::Vector{Int64}, subexpressions::Vector{Vector{MathOptInterface.Nonlinear.Node}}, subexpression_dependency_graph::Vector{Vector{Int64}})
    @ MathOptInterface.Nonlinear.ReverseAD ~/.julia/packages/MathOptInterface/LqT2k/src/Nonlinear/ReverseAD/graph_tools.jl:394
  [3] _order_subexpressions(main_expressions::Vector{Vector{MathOptInterface.Nonlinear.Node}}, subexpressions::Vector{Vector{MathOptInterface.Nonlinear.Node}})
    @ MathOptInterface.Nonlinear.ReverseAD ~/.julia/packages/MathOptInterface/LqT2k/src/Nonlinear/ReverseAD/graph_tools.jl:460
  [4] initialize(d::MathOptInterface.Nonlinear.ReverseAD.NLPEvaluator, requested_features::Vector{Symbol})
    @ MathOptInterface.Nonlinear.ReverseAD ~/.julia/packages/MathOptInterface/LqT2k/src/Nonlinear/ReverseAD/mathoptinterface_api.jl:50
  [5] initialize(evaluator::MathOptInterface.Nonlinear.Evaluator{MathOptInterface.Nonlinear.ReverseAD.NLPEvaluator}, features::Vector{Symbol})
    @ MathOptInterface.Nonlinear ~/.julia/packages/MathOptInterface/LqT2k/src/Nonlinear/evaluator.jl:89
  [6] parser_nonlinear_expression(cmodel::JuMP.Model, nvar::Int64, F::Vector{JuMP.NonlinearExpression}; hessian::Bool)
    @ NLPModelsJuMP ~/work/NLPModelsJuMP.jl/NLPModelsJuMP.jl/src/utils.jl:460
  [7] NLPModelsJuMP.MathOptNLSModel(cmodel::JuMP.Model, F::Vector{JuMP.NonlinearExpression}; hessian::Bool, name::String)
    @ NLPModelsJuMP ~/work/NLPModelsJuMP.jl/NLPModelsJuMP.jl/src/moi_nls_model.jl:28
  [8] hs70()
    @ NLSProblems ~/work/NLPModelsJuMP.jl/NLPModelsJuMP.jl/NLSProblems.jl/src/hs70.jl:64
  [9] top-level scope
    @ ~/work/NLPModelsJuMP.jl/NLPModelsJuMP.jl/NLSProblems.jl/test/runtests.jl:17
 [10] include(fname::String)
    @ Base.MainInclude ./client.jl:476
 [11] top-level scope
    @ none:6
in expression starting at /home/runner/work/NLPModelsJuMP.jl/NLPModelsJuMP.jl/NLSProblems.jl/test/runtests.jl:14
┌ Error: Pkg.Types.PkgError("Package NLSProblems errored during testing")
└ @ Main none:15
```

**Comment by tmigot** [comment]
and also a problem with `camshape` https://github.com/JuliaSmoothOptimizers/OptimizationProblems.jl/blob/main/src/PureJuMP/camshape.jl with the following mysterious:
```
ERROR: LoadError: Interval constraint contains non-constant left- or right-hand sides. Reformulate as two separate constraints, or move all variables into the central term.
Stacktrace:
 [1] error(s::String)
   @ Base ./error.jl:35
 [2] _normalize_constraint_expr(lhs::Expr, body::Expr, rhs::Expr)
   @ JuMP ~/.julia/packages/JuMP/bm7X3/src/nlp.jl:389
 [3] _expr_to_constraint(expr::Expr)
   @ JuMP ~/.julia/packages/JuMP/bm7X3/src/nlp.jl:404
 [4] add_nonlinear_constraint(model::JuMP.Model, ex::Expr)
   @ JuMP ~/.julia/packages/JuMP/bm7X3/src/nlp.jl:440
 [5] macro expansion
   @ ~/.julia/packages/JuMP/bm7X3/src/macros.jl:2353 [inlined]
 [6] camshape(; n::Int64, kwargs::Base.Pairs{Symbol, Union{}, Tuple{}, NamedTuple{(), Tuple{}}})
   @ OptimizationProblems.PureJuMP ~/work/NLPModelsJuMP.jl/NLPModelsJuMP.jl/OptimizationProblems.jl/src/PureJuMP/camshape.jl:26
 [7] top-level scope
   @ ~/work/NLPModelsJuMP.jl/NLPModelsJuMP.jl/OptimizationProblems.jl/test/runtests.jl:43
 [8] include(fname::String)
   @ Base.MainInclude ./client.jl:476
 [9] top-level scope
   @ none:6
in expression starting at /home/runner/work/NLPModelsJuMP.jl/NLPModelsJuMP.jl/OptimizationProblems.jl/test/runtests.jl:39
┌ Error: Pkg.Types.PkgError("Package OptimizationProblems errored during testing")
└ @ Main none:15
```

**Comment by tmigot** [comment]
> 

@amontoison  The indices in the linear API are relative indices.

File: src/moi_nls_model.jl [hunk 4]
```diff
@@ -268,8 +268,8 @@ function NLPModels.jac_nln_structure!(
   rows::AbstractVector{<:Integer},
   cols::AbstractVector{<:Integer},
 )
-  view(rows, nls.lincon.nnzj+1:nls.meta.nnzj) .= nls.nlcon.jac_rows
-  view(cols, nls.lincon.nnzj+1:nls.meta.nnzj) .= nls.nlcon.jac_cols
+  view(rows, 1:(nls.nlcon.nnzj)) .= nls.nlcon.jac_rows
```
Comment (amontoison) [inline src/moi_nls_model.jl:271]: I don't understand Tangi, the rows and cols of the non-linear jacobian should be stored at the positions `nls.lincon.nnzj+1:nls.meta.nnzj)`.

Comment (tmigot) [reply src/moi_nls_model.jl:271]: No, `jac_nln_structure` and `jac_nln_coord` returns the jacobian of the nonlinear constraints only. It doesn't update the nonlinear rows of the whole jacobian.

File: docs/make.jl [hunk 5]
```diff
@@ -7,7 +7,6 @@ makedocs(
   strict = true,
   format = Documenter.HTML(
     prettyurls = get(ENV, "CI", nothing) == "true",
-    ansicolor=true,
```
Comment (amontoison) [inline docs/make.jl]: We just added it :)

**Comment by amontoison** [comment]
> and also a problem with `camshape` https://github.com/JuliaSmoothOptimizers/OptimizationProblems.jl/blob/main/src/PureJuMP/camshape.jl with the following mysterious:
> 
> ```
> ERROR: LoadError: Interval constraint contains non-constant left- or right-hand sides. Reformulate as two separate constraints, or move all variables into the central term.
> Stacktrace:
>  [1] error(s::String)
>    @ Base ./error.jl:35
>  [2] _normalize_constraint_expr(lhs::Expr, body::Expr, rhs::Expr)
>    @ JuMP ~/.julia/packages/JuMP/bm7X3/src/nlp.jl:389
>  [3] _expr_to_constraint(expr::Expr)
>    @ JuMP ~/.julia/packages/JuMP/bm7X3/src/nlp.jl:404
>  [4] add_nonlinear_constraint(model::JuMP.Model, ex::Expr)
>    @ JuMP ~/.julia/packages/JuMP/bm7X3/src/nlp.jl:440
>  [5] macro expansion
>    @ ~/.julia/packages/JuMP/bm7X3/src/macros.jl:2353 [inlined]
>  [6] camshape(; n::Int64, kwargs::Base.Pairs{Symbol, Union{}, Tuple{}, NamedTuple{(), Tuple{}}})
>    @ OptimizationProblems.PureJuMP ~/work/NLPModelsJuMP.jl/NLPModelsJuMP.jl/OptimizationProblems.jl/src/PureJuMP/camshape.jl:26
>  [7] top-level scope
>    @ ~/work/NLPModelsJuMP.jl/NLPModelsJuMP.jl/OptimizationProblems.jl/test/runtests.jl:43
>  [8] include(fname::String)
>    @ Base.MainInclude ./client.jl:476
>  [9] top-level scope
>    @ none:6
> in expression starting at /home/runner/work/NLPModelsJuMP.jl/NLPModelsJuMP.jl/OptimizationProblems.jl/test/runtests.jl:39
> ┌ Error: Pkg.Types.PkgError("Package OptimizationProblems errored during testing")
> └ @ Main none:15
> ```

We should share it with @odow. The problem seems to be related to JuMP directly.

**Comment by odow** [comment]
I have a fix: https://github.com/jump-dev/JuMP.jl/pull/3045. I'll tag v1.2.1 asap.

File: docs/make.jl (hunk 5 above)
Comment (tmigot) [reply docs/make.jl]: I rebased

**Comment by tmigot** [comment]
Thanks @odow for the fix. The tests now pass on OptimizationProblems.jl

**Comment by tmigot** [comment]
The second problem with `hs70` on `NLSProblems.jl` seems to be on our side. Here is a MWE of our process:
```
using NLPModels, NLPModelsJuMP, JuMP

nls = Model()
x0 = [2; 4; 0.04; 2]
uvar = [100; 100; 1; 100]
@variable(nls, 0.00001 ≤ x[i = 1:4] ≤ uvar[i], start = x0[i])

c = [(i ≥ 2) ? i - 1 : 0.1 for i = 1:19]
yobs = [
    0.00189
    0.1038
    0.268
    0.506
    0.577
    0.604
    0.725
    0.898
    0.947
    0.845
    0.702
    0.528
    0.385
    0.257
    0.159
    0.0869
    0.0453
    0.01509
    0.00189
  ]

@NLexpression(nls, b, x[3] + (1 - x[3]) * x[4])
@NLexpression(
    nls,
    ycal[i = 1:19],
    (1 + 1 / (12 * x[2])) *
    (x[3] * b^x[2]) *
    ((x[2] / 6.2832)^(0.5)) *
    (c[i] / 7.685)^(x[2] - 1) *
    exp(x[2] - b * c[i] * x[2] / 7.658) +
    (1 + (1 / (12 * x[1]))) *
    (1 - x[3]) *
    (b / x[4])^x[1] *
    (x[1] / 6.2832)^0.5 *
    (c[i] / 7.658)^(x[1] - 1) *
    exp(x[1] - b * c[i] * x[1] / (7.658 * x[4]))
  )

@NLexpression(nls, F[i = 1:19], ycal[i] - yobs[i])

Fmodel = JuMP.Model()
@variable(Fmodel, x[1:4])
JuMP._init_NLP(Fmodel)
@objective(Fmodel, Min, 0.0)

@NLconstraint(Fmodel, F[1] == 0)
@NLconstraint(Fmodel, F[2] == 0)
@NLconstraint(Fmodel, F[3] == 0)

Feval = NLPEvaluator(Fmodel)
MOI.initialize(Feval, [:Grad, :Jac, :JacVec, :Hess, :HessVec]) # fails as `d.data.expressions` is empty here https://github.com/jump-dev/MathOptInterface.jl/blob/fd8ca78e625f4e4f3fbaa94841491cea431d1336/src/Nonlinear/ReverseAD/mathoptinterface_api.jl#L52
```
I believe the fix would be to copy the `NLExpressions` in the `FModel` as the following works:
```
Fmodel = JuMP.Model()
@variable(Fmodel, x[1:4])
JuMP._init_NLP(Fmodel)
@objective(Fmodel, Min, 0.0)

@NLexpression(Fmodel, b, x[3] + (1 - x[3]) * x[4])
@NLexpression(
    Fmodel,
    ycal[i = 1:19],
    (1 + 1 / (12 * x[2])) *
    (x[3] * b^x[2]) *
    ((x[2] / 6.2832)^(0.5)) *
    (c[i] / 7.685)^(x[2] - 1) *
    exp(x[2] - b * c[i] * x[2] / 7.658) +
    (1 + (1 / (12 * x[1]))) *
    (1 - x[3]) *
    (b / x[4])^x[1] *
    (x[1] / 6.2832)^0.5 *
    (c[i] / 7.658)^(x[1] - 1) *
    exp(x[1] - b * c[i] * x[1] / (7.658 * x[4]))
  )
@NLexpression(Fmodel, F[i = 1:19], ycal[i] - yobs[i])

@NLconstraint(Fmodel, F[1] == 0)
@NLconstraint(Fmodel, F[2] == 0)
@NLconstraint(Fmodel, F[3] == 0)

Feval = NLPEvaluator(Fmodel)
MOI.initialize(Feval, [:Grad, :Jac, :JacVec, :Hess, :HessVec])
```

**Comment by odow** [comment]
You can't use an expression (or variable) defined in one model inside another model. 

What is the goal you're trying to achieve?

**Comment by amontoison** [comment]
> You can't use an expression (or variable) defined in one model inside another model.
> 
> What is the goal you're trying to achieve?

@odow We have a nonlinear least-squares model with constraints (`cmodel`) and we create a second model (`Fmodel`) that only models the constraints `Fi = 0` where the `Fi` are expressions that composed the objective of the first model (`Σ Fi^2`).

**Comment by tmigot** [comment]
@amontoison I added the constraint also when `Fi` is an Array of expression that was the missing key. I also took the liberty to re-write it using dispatch :). Now, the unit tests pass and the breakage as well!

**Review by tmigot — APPROVED** [review]

**Comment by abelsiqueira** [comment]
Thanks for doing the harder part, guys. I am merging this.

**Comment by odow** [comment]
Can you open an issue? Thats unexpected. I'm away from my computer so will
look next week.

**Pull request merged**